File: medium/0200_number_of_islands/solution.cpp
#include <stack>
#include <vector>

class Solution {
public:
  int numIslands(std::vector<std::vector<char>> &grid) {
    const size_t rows = grid.size();
    const size_t cols = grid[0].size();

    int numIslands = 0;

    for (size_t row{}; row < rows; ++row) {
      for (size_t col{}; col < cols; ++col) {
        if (grid[row][col] == '1') {
          std::stack<std::pair<size_t, size_t>> s;
          s.push({row, col});

          while (!s.empty()) {
            auto [r, c] = s.top();
            s.pop();
            grid[r][c] = '0';

            if (r - 1 != -1 && grid[r - 1][c] == '1')
              s.push({r - 1, c});
            if (c + 1 < cols && grid[r][c + 1] == '1')
              s.push({r, c + 1});
            if (r + 1 < rows && grid[r + 1][c] == '1')
              s.push({r + 1, c});
            if (c - 1 != -1 && grid[r][c - 1] == '1')
              s.push({r, c - 1});
          }

          ++numIslands;
        }
      }
    }

    return numIslands;
  }
};

File: medium/0200_number_of_islands/solution.py
class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:
        rows, cols = len(grid), len(grid[0])
        num_islands = 0

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1":
                    stack = {(r, c)}

                    while stack:
                        rr, cc = stack.pop()

                        grid[rr][cc] = "0"

                        if rr + 1 < rows and grid[rr + 1][cc] == "1":
                            stack.add((rr + 1, cc))
                        if cc + 1 < cols and grid[rr][cc + 1] == "1":
                            stack.add((rr, cc + 1))
                        if rr - 1 >= 0 and grid[rr - 1][cc] == "1":
                            stack.add((rr - 1, cc))
                        if cc - 1 >= 0 and grid[rr][cc - 1] == "1":
                            stack.add((rr, cc - 1))

                    num_islands += 1

        return num_islands
